modules/python/python-stats.c

```diff
@@ -0,0 +1,206 @@
+
+      stats_unlock();
+    }
+  return (PyObject *) self;
+};
+
+static void
+py_stats_free(PyStats *self)
+{
+  stats_lock();
+  StatsClusterKey sc_key;
+  stats_cluster_single_key_set(&sc_key, SCS_PYTHON, self->stats_name, NULL);
+  stats_unregister_counter(&sc_key, SC_TYPE_SINGLE_VALUE, &(self->counter));
+  stats_unlock();
+
+  g_free( (void *) self->stats_name);
+
+  PyObject_Del(self);
+}
+
+static PyObject *
+py_stats_increase(PyStats *self, PyObject *unused)
+{
+  stats_counter_inc(self->counter);
+  Py_RETURN_NONE;
+}
+
+static PyObject *
+py_stats_decrease(PyStats *self, PyObject *unused)
+{
+  stats_counter_dec(self->counter);
+  Py_RETURN_NONE;
+}
+
+static PyObject *
+py_stats_add(PyStats *self, PyObject *args)
+{
+  gssize amount;
+  if (!PyArg_ParseTuple(args,"n", &amount))
+    {
+      return NULL;
+    }
+  stats_counter_add(self->counter, amount);
+  Py_RETURN_NONE;
+}
+
+static PyObject *
+py_stats_sub(PyStats *self, PyObject *args)
+{
+  gssize amount;
+  if (!PyArg_ParseTuple(args,"n", &amount))
+    {
+      return NULL;
+    }
+
+  stats_counter_sub(self->counter, amount);
+  Py_RETURN_NONE;
+}
+
+static PyMethodDef py_stats_methods[] =
+{
+  {
+    "increase", (PyCFunction)py_stats_increase,
+    METH_NOARGS, "Increase counter"
+  },
+  {
+    "decrease", (PyCFunction)py_stats_decrease,
+    METH_NOARGS, "Decrease counter"
+  },
+  {
+    "add", (PyCFunction)py_stats_add,
+    METH_VARARGS, "Increase counter with amount"
+  },
+  {
+    "sub", (PyCFunction)py_stats_sub,
+    METH_VARARGS, "Decrease counter with amount"
+  },
+  {NULL}
+};
+
+static PyTypeObject py_stats_type =
+{
+  PyVarObject_HEAD_INIT(&PyType_Type, 0)
+  .tp_name = "StatsCounter",
+  .tp_basicsize = sizeof(PyStats),
+  .tp_dealloc = (destructor) py_stats_free,
+  .tp_flags = Py_TPFLAGS_DEFAULT | Py_TPFLAGS_BASETYPE,
+  .tp_doc = "StatsCounter representing a counter in syslog-ng statistics",
+  .tp_new = py_stats_new,
+  .tp_methods = py_stats_methods,
+  .tp_str = NULL,
+  .tp_as_mapping = NULL,
+  0,
+};
+
+static PyMethodDef stats_module_methods[] =
+{
+  {NULL}
+};
+
+#if PY_MAJOR_VERSION >= 3
+static struct PyModuleDef moduledef =
+{
+  PyModuleDef_HEAD_INIT,
+  "sngstats",     /* m_name */
+  "Module for encapsulating StatsCounter type",  /* m_doc */
+  -1,                  /* m_size */
+  stats_module_methods,    /* m_methods */
+  NULL,                /* m_reload */
+  NULL,                /* m_traverse */
+  NULL,                /* m_clear */
+  NULL,                /* m_free */
+};
+#endif
+
+static PyObject *
+py_create_stats_module(void)
+{
+  PyObject *m;
+  if (PyType_Ready(&py_stats_type) < 0)
+    return NULL;
+
+#if PY_MAJOR_VERSION >= 3
+  m = PyModule_Create(&moduledef);
+#else
+  m = Py_InitModule3("sngstats", stats_module_methods,
+                     "Module for encapsulating StatsCounter type");
+#endif
+
+  Py_INCREF((PyObject *) &py_stats_type);
+  PyModule_AddObject(m, "StatsCounter", (PyObject *) &py_stats_type);
```

Comment:
Just a hint but
https://docs.python.org/3/c-api/import.html#c.PyImport_AddModule
returns with a newly created python module if it does not exist already. According to documentation, it works both with python2 and python3. So you could avoid the complex module registration altogether, branching over python2 and python3, all you need is a name for the new module. As I see you create an "empty" module, the only specialization is the docstring, which would apparently be lost with my suggested version, but the patch would be much simpler, and not need to touch compat.

The suggested line: 
```
PyModule_AddObject(PyImport_AddModule("syslogng"), "StatsCounter", (PyObject *) &py_stats_type);
```